source/common/quic/envoy_tls_server_handshaker.cc

```diff
@@ -0,0 +1,115 @@
+    const std::string& hostname, absl::string_view client_hello, const std::string& alpn,
+    std::optional<std::string> alps, const std::vector<uint8_t>& quic_transport_params,
+    const std::optional<std::vector<uint8_t>>& early_data_context,
+    const quic::QuicSSLConfig& ssl_config) {
+  return default_proof_source_handle_->SelectCertificate(
+    server_address, client_address, original_connection_id, ssl_capabilities,
+    hostname, client_hello, alpn, alps, quic_transport_params, early_data_context, ssl_config);
+}
+
+quic::QuicAsyncStatus EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::ComputeSignature(
+    const quic::QuicSocketAddress& server_address, const quic::QuicSocketAddress& client_address,
+    const std::string& hostname, uint16_t signature_algorithm, absl::string_view in,
+    size_t max_signature_size) {
+  auto res = proof_source_->getTransportSocketAndFilterChain(server_address, client_address, hostname);
+  if (!res.has_value()) {
+    return quic::QUIC_FAILURE;
+  }
+
+  private_key_method_provider_ =
+      res->transport_socket_factory_.getPrivateKeyMethodProvider(hostname);
+    
+  if (private_key_method_provider_ == nullptr) {
+    return default_proof_source_handle_->ComputeSignature(server_address,
+      client_address, hostname, signature_algorithm, in, max_signature_size);
+  } else {
+    filter_chain_ = res->filter_chain_;
+    auto* ssl = handshaker_->GetSsl();
+    private_key_method_provider_->registerPrivateKeyMethod(ssl, *this, handshaker_->dispatcher());
+
+    max_sig_size_ = max_signature_size;
+    signature_.resize(max_sig_size_);
+    size_t sig_size = 0;
+
+    auto ret = private_key_method_provider_->getBoringSslPrivateKeyMethod()->sign(
+        ssl, const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(signature_.data())), &sig_size,
+        max_sig_size_, signature_algorithm,
+        const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(in.data())), in.size());
+
+    if (ret == ssl_private_key_failure) {
+      return quic::QUIC_FAILURE;
+    }
+
+    if (ret == ssl_private_key_retry) {
+      return quic::QUIC_PENDING;
+    }
+
+    return quic::QUIC_SUCCESS;
+  }
+}
+
+void EnvoyTlsServerHandshaker::EnvoyDefaultProofSourceHandle::onPrivateKeyMethodComplete() {
+  ASSERT(private_key_method_provider_ != nullptr);
+  ASSERT(filter_chain_ != absl::nullopt);
+  size_t sig_len = 0;
+  auto ret = private_key_method_provider_->getBoringSslPrivateKeyMethod()->complete(
+      handshaker_->GetSsl(), const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(signature_.data())), &sig_len,
+      max_sig_size_);
+  if (ret == ssl_private_key_failure) {
+    handshaker_->OnComputeSignatureDone(
+      false, false, "", nullptr);
+  } else { 
+    ASSERT(filter_chain_ != absl::nullopt);
+    handshaker_->OnComputeSignatureDone(
+                true, false, std::move(signature_), std::make_unique<EnvoyQuicProofSourceDetails>(filter_chain_.value().get()));
```

Comment:
I suspect you don't need to do the same thing as EnvoyQuicProofSource does to supply an EnvoyQuicProofSourceDetails object. This is specific to EnvoyQuicProofSource. As a result you also don't have to pass around proof_source_ and filter_chain_, all you need to pass around is the transport socket factory I noted in the other comment.

Reply:
I got a crash when pass a nullptr for the `EnvoyQuicProofSourceDetails`. but yes, I didn't track down where need it, but after I pass down `EnvoyQuicProofSourceDetails`, the crash is gone.

Reply:
Emm...I was wrong, it seems EnvoyQuicProofSourceDetails is not used at anywhere. It should be fine to just pass a nullptr.